app/jobs/maintenance_tasks/task_job.rb

```diff
@@ -35,8 +35,13 @@ def build_enumerator(_run, cursor:)
       when Array
         enumerator_builder.build_array_enumerator(collection, cursor: cursor)
       else
-        raise ArgumentError, "#{@task.class.name}#collection must be either "\
-          'an Active Record Relation or an Array.'
+        if dynamic_collection?(collection)
+          collection.call(cursor: cursor)
```

Comment:
It just occurred to me while exploring alternatives that this wouldn't address the case where we want to `enumerator_builder.active_record_on_batches`, since we don't have a reference to `enumerator_builder`.

I wonder if instead it would make sense to set `@cursor` on the job, and then inject the entire job in, so you can something like

```ruby
def collection
  lambda do |job:|
    job.enumerator_builder.active_record_on_batches(
      Product.all,
      cursor: job.cursor,
      batch_size: 100,
    )
  end
end
```

This might even be useful for future use-cases like run-specific params 🤔 

```ruby
def collection
  lambda do |job:|
    job.enumerator_builder.active_record_on_batches(
      Product.where(shop_id: job.arguments.second[:shop_id]),
      cursor: job.cursor,
      batch_size: 100,
    )
  end
end
```

Reply:
This is leaking too much of the implementation to the users in my option. People should not even know a cursor exists, or an enumerator builder. What people want is to be able to create batches on in the collection, so maybe we should use a declarative approach in the task like:

```
run_in_barches_of 100

def collection
  Product.all
end
```

Reply:
One of the questions I struggled with when designing this was

> How to accommodate complex use cases without adding complexity to the simple ones?

I see what you mean, but I'd argue that in the complex use case the author already has to be aware of the cursor (and why it's needed) and it's not that much more complexity to be aware of the enumerator builder. We could always pass in some sort of simpler object than the entire job, like

```ruby
CollectionContext = Struct.new(:cursor, :enumerator_builder, keyword_init: true)
```

which would easily allow future expansion, such as forwarding run-specific parameters. Alternatively, we just make a breaking change if the time comes. 🤷‍♂️ 

The `run_in_batches_of` idea is intriguing though.

----

Another thing I noticed in the [implementation of `JobIteration::EnumeratorBuilder`](https://github.com/Shopify/job-iteration/blob/ae9f11e316c99eaf0c5a478c1fa0a659fcd7730a/lib/job-iteration/enumerator_builder.rb#L11-L27) is that there may be plans to force enumerators to be wrapped in a `JobIteration::EnumeratorBuilder::Wrapper`:

```ruby
    # These wrappers ensure we have a custom type that we can assert on in
    # Iteration. It's useful that the `wrapper` passed to EnumeratorBuilder in
    # `enumerator_builder` is _always_ the type that is returned from
    # `build_enumerator`. This prevents people from implementing custom
    # Enumerators without wrapping them in
    # `enumerator_builder.wrap(custom_enum)`. We don't do this yet for backwards
    # compatibility with raw calls to EnumeratorBuilder. Think of these wrappers
    # the way you should a middleware.
    class Wrapper < Enumerator
      def self.wrap(_builder, enum)
        # ...
```

If this comes to fruition, it may force the custom enumerator use case to be aware of the `enumerator_builder` so it can call `.wrap` on it. 🤔 Or maybe we just auto-wrap if needed.

Reply:
One of the design philosophy of this library is make complex use cases easy to do. We don't aim to give all control and power to the users so they can build complex code in their apps. The code written to use this app should be easy to read and complex things should be made simple with less code possible.

In other words, we should focus our energy to implement something that is easy to use, not in allowing people to have access to all the nobs and implementation detail possible.

I think the fact that anyone need to now about this cursor is a huge design flaw in the job_iterator library. People usually don't do anything with this cursor, they don't define anything different, they just pass around. If people always pass around the same thing, why do we make them pass around anyway?

I don't want this library to make the same mistakes. The API needs to follow more conventions, use less configurations, make complex cases easier.

So maybe instead of accepting this PR as is I should be asking, is there anything we can make easier  in this library so you don't need to write a enumerator for your task? The CSV use case was one of them.

Reply:
Thanks for the context and thorough review @rafaelfranca 👍

> is there anything we can make easier in this library so you don't need to write an enumerator for your task?

I have added some more concrete example use cases to https://github.com/Shopify/maintenance_tasks/issues/207#issuecomment-764189242. Ultimately though, they're similar to those described in [`JobIteration`'s _Custom Enumerator Guide_](https://github.com/Shopify/job-iteration/blob/master/guides/custom-enumerator.md): the application needs to iterate over a resource type the iteration framework should not know about.

I spent some time thinking about this (maybe too much), and think I figured out why it's so tricky: `JobIteration`'s `build_enumerator` requires knowledge about _how to build the enumerator_ **and** _how far along it is in the iteration_. If instead, we could build an object that knows how to iterate, but defer telling it how far it is in the iteration, then we could encapsulate the `cursor`.

It's really rough, but I've tried to illustrate my idea in [this Gist](https://gist.github.com/sambostock/3323f401fd3061f1f8e620fa92b44da1), the key being basically this change:

```diff
 def interruptible_perform(*arguments)
    # ...
-   enumerator = build_enumerator(*arguments, cursor: cursor_position)
+   enumerator = build_enumerator(*arguments).call(cursor: cursor_position)
    # ...
 end
```

**If we did that, only maintainers and custom enumerator authors would need to know about `cursor`, not consumers using existing enumerators.**

Of course... that would likely be breaking changes to both `JobIteration` and `MaintenanceTasks`... 😞